virtualization-and-containers/material/docker_slides.md

```diff
@@ -0,0 +1,252 @@
+- Containers for consistent development environment
+- Containers for consistent testing environment
+- Portable format for sharing applications
+- Avoid tedious installation procedures by providing Docker container ([FEniCS](https://fenicsproject.org/download/), [GitLab](https://docs.gitlab.com/ee/install/docker.html)...)
+
+---
+
+## Building Blocks 1/2
+
+- Docker daemon `dockerd`
+  - Controlling instance of containers and reacts to API requests
+  - Server process
+- Docker client
+  - User interface/tools to interact with, create, manage containers etc. via daemon
+- Docker registries
+  - Registries that manage Docker images to be used
+
+---
+
+## Building Blocks 2/2
+
+- Docker objects
+  - **Images**
+    - Read-only template for creating a container
+    - An image can be based on another image
+  - **Containers**
+    - Runnable instance of an image
+
+---
+
+## Docker Architecture
+
+<img src="https://docs.docker.com/engine/images/architecture.svg" width=80%; style="margin-left:auto; margin-right:auto; padding-top: 25px; padding-bottom: 25px">
+
+[https://docs.docker.com/engine/images/architecture.svg](https://docs.docker.com/engine/images/architecture.svg)
+
+---
+
+## Connection to Host
+
+- Container communicates via daemon `dockerd` (runs as   root)
+- Strong isolation (`namespaces` and `cgroups`)
+  - You cannot access Host filesystem by default
+  - Several [mount options](https://docs.docker.com/storage) available
+
+---
+
+## Requirements
+
+- Root rights for installation
+- `dockerd` runs as root -> Interaction needs root rights
+  - Prefix commands with `sudo`
+  - Be member of group `docker` (=makes you root), expected by some applications (e.g. `act`)
+  - [Attack surface?!](https://docs.docker.com/engine/security/#docker-daemon-attack-surface)
+    - [Isolate user namespace](https://docs.docker.com/engine/security/userns-remap/)
+    - Use trustworthy containers
+- Alternatives:
+  - [Rootless mode](https://docs.docker.com/engine/security/rootless/)
+  - Run Docker in a VM
+- Check [security notes](https://docs.docker.com/engine/security/)
+
+---
+
+## Useful Commands 1/2
+
+- `docker run OPTIONS`
+  - Run a container
+- `docker image ls`
+  - List locally available images
+- `docker pull NAME:TAG`
+  - Pulls an image from registry, `TAG` optional
+- `docker container create IMAGE`
+  - Create container from image
+- `docker container ls`
+  - List running containers
+  - Add `-a` to see all containers
+
+---
+
+## Useful Commands 2/2
+
+- `docker container start/stop NAME`
+  - Start/stop container
+- `docker container attach NAME`
+  - Attach to running container
+- `docker build`
+  - Creates an image from a given Dockerfile
+- `docker cp`
+  - Copy files in/out of container
+- `docker image history IMAGE`
+  - Show layers of image (including commands)
+- `docker system prune`
+  - Remove all unused objects (images, containers...)
+- Many commands have similar/same outcome
+
+---
+
+## Demo: Running prebuilt images
```

Comment:
Similar to above, it would be nice to have your notes somehow available

Reply:
Notes will be in the `overview.md`.